apps/ejabberd/src/ejabberd_s2s_in.erl

```diff
@@ -202,40 +202,28 @@ wait_for_stream({xmlstreamstart, _Name, Attrs}, StateData) ->
-                               StateData#state.socket) of
-                            {ok, Cert} ->
-                                case (StateData#state.sockmod):get_verify_result(StateData#state.socket) of
-                                    0 ->
-                                        [#xmlel{name = <<"mechanisms">>,
-                                                attrs = [{<<"xmlns">>, ?NS_SASL}],
-                                                children = [#xmlel{name = <<"mechanism">>,
-                                                                   children = [#xmlcdata{content = <<"EXTERNAL">>}]}]}];
-                                    CertVerifyRes ->
-                                        case StateData#state.tls_certverify of
-                                            true -> {error_cert_verif, CertVerifyRes, Cert};
-                                            false -> []
-                                        end
-                                end;
-                            error ->
-                                []
-                        end;
+                case StateData#state.tls_enabled of
                     true ->
+                        get_sasl(StateData#state.sockmod,
+                                 StateData#state.socket,
+                                 StateData#state.tls_certverify);
+                    _Else ->
                         []
                 end,
-            StartTLS = if
-                           StateData#state.tls_enabled ->
-                               [];
-                           (not StateData#state.tls_enabled) and (not StateData#state.tls_required) ->
-                              [#xmlel{name = <<"starttls">>,
-                                      attrs = [{<<"xmlns">>, ?NS_TLS}]}];
-                           (not StateData#state.tls_enabled) and StateData#state.tls_required ->
-                              [#xmlel{name = <<"starttls">>,
-                                      attrs = [{<<"xmlns">>, ?NS_TLS}],
-                                      children = [#xmlel{name = <<"required">>}]}]
-                       end,
+            StartTLS =
+                case StateData of
+                    StateData when StateData#state.tls_enabled ->
+                        [];
+                    StateData when (not StateData#state.tls_enabled) and
+                                   (not StateData#state.tls_required) ->
+                        [#xmlel{name = <<"starttls">>,
+                                attrs = [{<<"xmlns">>, ?NS_TLS}]}];
+                    StateData when (not StateData#state.tls_enabled) and
+                                   StateData#state.tls_required ->
+                        [#xmlel{name = <<"starttls">>,
+                                attrs = [{<<"xmlns">>, ?NS_TLS}],
+                                children = [#xmlel{name = <<"required">>}]}]
+                end,
```

Comment:
Same as above. It is legacy code and it is horrible, we try to improve it whenever possible, this is a good opportunity do to so.